### PR TITLE
Rewrite functionality that gets latest package versions for package owners

### PR DESCRIPTION
**Pull request by xavierdecoster**
See [this comment](https://github.com/NuGet/NuGetGallery/issues/4782#issuecomment-359814380)

### CODE REVIEW

File: src/NuGetGallery/Services/PackageService.cs [hunk 1]
```diff
@@ -216,123 +216,80 @@ public virtual Package FindAbsoluteLatestPackageById(string id, int? semVerLevel
...
+                if (latest != null)
                 {
-                    latestPackage = latestPackagesById.First();
+                    yield return latest;
+                    continue;
                 }
-                else
+
+                // Return latest stable non-SemVer2 version if available
+                latest = packagesByRegistration.Where(p => p.IsLatestStable).FirstOrDefault();
```
Comment (joelverhagen) [inline src/NuGetGallery/Services/PackageService.cs]: How could you have a latest stable non-SemVer2 but no latest stable SemVer2? SemVer2-compatible (i.e. SemVer 1.0.0 and SemVer 2.0.0) is a superset of non-SemVer2 (i.e. just SemVer 1.0.0)

File: src/NuGetGallery/Services/PackageService.cs [hunk 2]
```diff
@@ -216,123 +216,80 @@ public virtual Package FindAbsoluteLatestPackageById(string id, int? semVerLevel
...
+                if (latest != null)
                 {
-                    mergedResults.Add(latestPackage.PackageRegistration.Id, latestPackage);
+                    yield return latest;
+                    continue;
                 }
+
+                // Like DisplayPackage we should prefer to show you information from the latest stable version,
+                // but show you the latest version (potentially latest UNLISTED version) otherwise.
+                yield return packagesByRegistration.OrderByDescending(p => p.Version).First();
```
Comment (joelverhagen) [inline src/NuGetGallery/Services/PackageService.cs]: Is `p.Version` a `string` or a `NuGetVersion`?

Comment (xavierdecoster) [reply src/NuGetGallery/Services/PackageService.cs]: It's the `string Version` property on the `Package` entity. I sense this might need to be parsed as a `NuGetVersion` before ordering :)

Comment (joelverhagen) [reply src/NuGetGallery/Services/PackageService.cs]: Haha, yup!

File: src/NuGetGallery/Services/PackageService.cs (hunk 1 above)
Comment (xavierdecoster) [reply src/NuGetGallery/Services/PackageService.cs]: It shouldn't happen, but just in case, by taking this "fallback" approach, the code is a bit more resilient against possible concurrency issues.

File: src/NuGetGallery/Services/PackageService.cs (hunk 2 above)
Comment (xavierdecoster) [reply src/NuGetGallery/Services/PackageService.cs]: Fixed. Good catch!

File: src/NuGetGallery/Services/PackageService.cs (hunk 1 above)
Comment (joelverhagen) [reply src/NuGetGallery/Services/PackageService.cs]: Alright, makes sense.

File: src/NuGetGallery/Services/PackageService.cs [hunk 3]
```diff
@@ -216,123 +216,80 @@ public virtual Package FindAbsoluteLatestPackageById(string id, int? semVerLevel
...
             {
-                Package latestPackage;
-                if (includeUnlisted)
+                // Return latest stable SemVer2 version if available
+                var latest = packagesByRegistration.Where(p => p.IsLatestStableSemVer2).FirstOrDefault();
+                if (latest != null)
                 {
-                    latestPackage = latestPackagesById.First();
+                    yield return latest;
+                    continue;
```
Comment (cristinamanum) [inline src/NuGetGallery/Services/PackageService.cs]: Why is the continue needed?

File: src/NuGetGallery/Services/PackageService.cs [hunk 4]
```diff
@@ -216,123 +216,80 @@ public virtual Package FindAbsoluteLatestPackageById(string id, int? semVerLevel
...
-
-                mergedResults[latestStablePackage.PackageRegistration.Id] = latestStablePackage;
-            }
-        }
-
-        /// <summary>
-        /// Merge latest packages by owner, including organization owners that the user belongs to.
-        /// </summary>
-        private void MergeLatestPackagesByOwner(List<int> ownerKeys, bool includeUnlisted, Dictionary<string, Package> mergedResults)
+        private IEnumerable<Package> GetLatestPackagesForOwners(IReadOnlyCollection<Package> packages)
```
Comment (cristinamanum) [inline src/NuGetGallery/Services/PackageService.cs]: This method is not related with an user, would a a method name without be better?

File: src/NuGetGallery/Services/PackageService.cs (hunk 3 above)
Comment (cristinamanum) [reply src/NuGetGallery/Services/PackageService.cs]: I think we would need unit tests to verify the logic described in the issue.

Comment (xavierdecoster) [reply src/NuGetGallery/Services/PackageService.cs]: The `continue` is needed to exit the current iteration in the `foreach` loop and return only a single package per id.

Comment (xavierdecoster) [reply src/NuGetGallery/Services/PackageService.cs]: We have existing unit tests covering this (current) logic.

File: src/NuGetGallery/Services/PackageService.cs (hunk 4 above)
Comment (xavierdecoster) [reply src/NuGetGallery/Services/PackageService.cs]: Renamed. It used to depend on the owners list, but forgot to rename after removing that parameter.
